lib/vrl/stdlib/src/hmac.rs

```diff
@@ -0,0 +1,243 @@
+use ::value::Value;
+use base64::Engine;
+use hmac::{Hmac as HmacHasher, Mac};
+use sha1::Sha1;
+use sha_2::{Sha224, Sha256, Sha384, Sha512};
+use vrl::prelude::*;
+
+macro_rules! hmac {
+    ($algorithm:ty, $key:expr, $val:expr) => {{
+        let mut mac =
+            <HmacHasher<$algorithm>>::new_from_slice($key.as_ref()).expect("key is bytes");
+        mac.update($val.as_ref());
+        let result = mac.finalize();
+        let code_bytes = result.into_bytes();
+        code_bytes.to_vec()
+    }};
+}
+
+fn hmac(value: Value, key: Value, algorithm: Value, encoding: Value) -> Resolved {
+    let value = value.try_bytes()?;
+    let key = key.try_bytes()?;
+    let algorithm = algorithm.try_bytes_utf8_lossy()?.as_ref().to_uppercase();
+    let encoding = encoding.try_bytes_utf8_lossy()?.as_ref().to_lowercase();
+
+    let code_bytes = match algorithm.as_str() {
+        "SHA1" => hmac!(Sha1, key, value),
+        "SHA-224" => hmac!(Sha224, key, value),
+        "SHA-256" => hmac!(Sha256, key, value),
+        "SHA-384" => hmac!(Sha384, key, value),
+        "SHA-512" => hmac!(Sha512, key, value),
+        _ => return Err(format!("Invalid algorithm: {}", algorithm).into()),
+    };
+
+    let hash = match encoding.as_str() {
+        "hex" => hex::encode(code_bytes),
+        "base64" => base64::engine::general_purpose::STANDARD.encode(code_bytes),
+        _ => return Err(format!("Invalid encoding: {}", encoding).into()),
+    };
```

Comment:
Is there a reason this needs to do the encoding within this function instead of just returning bytes and calling `encode_base16` or `encode_base64` on the result?

Reply:
Mostly opinion here, but it would be an unnecessary, and not readily-apparent, additional step an end-user must take in VRL to separately call `encode_*` on the resulting bytes to get back a usable string. An HMAC is not useful in any other form other than as encoded bytes. Why make this function perform only half the job?

Reply:
I could be wrong, but I think whether or not the bytes need to be encoded depend on how you are using them? If you are just storing the HMAC somewhere that can handle it as raw bytes, there is no need to encode it.

I tend to agree that we should separate these operations since otherwise we need to remember to add each binary encoding we add to VRL, also to here, and anywhere else we extend this pattern to.  The `encrypt_*` functions are similar in that they just return bytes that can be encoded if desired so not doing it here would at least be consistent within the VRL standard library.

I agree that we should make this behavior clear in the docs though.

Reply:
At a certain point, I throw in the towel. Most other languages' implementations of HMAC functions I've seen in the wild directly include encode/decode options, and only hex or base64. But I don't hold the commit keys to the kingdom. :) Pushed up another commit removing encoding functionality and made resultant major changes to the function's cue file.

Reply:
😄 I appreciate the flexibility here. I think returning raw bytes, for now, will maintain consistency within VRL for functions that return bytes, but this is an enhancement we could always add in later if it does turn out to be a common expectation from users (potentially across all functions that return "bytes").